lib/abbreviato/truncated_sax_document.rb

```diff
@@ -0,0 +1,198 @@
+      str = @html_coder.encode(truncate_string(decoded_string), :named)
+      str << tail if @something_has_been_truncated && remaining_length >= 0
+      str
+    else
+      encoded_string
+    end
+    append_to_truncated_string string_to_append, string_to_append.bytesize
+  end
+
+  # This method is called when the parser encounters an comment
+  def comment(string)
+    comment = comment_tag(string)
+    append_to_truncated_string(comment) if comment.bytesize <= remaining_length
+  end
+
+  # This method is called when the parser encounters cdata. In practice, this also
+  # gets called for this style of comment inside an element:
+  #
+  #   <style><!--
+  #     /* Font Definitions */
+  #     @font-face
+  #       {font-family:Wingdings;
+  #       panose-1:5 0 0 0 0 0 0 0 0 0;}
+  #   --></style>
+  #
+  def cdata_block(string)
+    append_to_truncated_string(string) if string.bytesize <= remaining_length
+  end
+
+  # This method is called when the parser encounters a closing tag
+  def end_element(name)
+    if ignore_mode?
+      exit_ignored_level
+      return
+    end
+
+    return if @max_length_reached || ignorable_tag?(name) || ignore_mode?
+
+    unless single_tag_element? name
+      @closing_tags.pop
+      # Don't count the length when closing a tag - it was accommodated when
+      # the tag was opened
+      append_to_truncated_string closing_tag(name), 0
+    end
+  end
+
+  def end_document
+    @closing_tags.reverse.each { |name| append_to_truncated_string closing_tag name }
+  end
+
+  private
+
+  def opening_tag(name, attributes)
+    attributes_string = attributes_to_string attributes
+    if single_tag_element? name
+      "<#{name}#{attributes_string}/>"
+    else
+      "<#{name}#{attributes_string}>"
+    end
+  end
+
+  def comment_tag(comment)
+    "<!--#{comment}-->"
+  end
+
+  def closing_tag(name)
+    "</#{name}>"
+  end
+
+  def remaining_length
+    max_length - estimated_length_with_tail
+  end
+
+  def estimated_length_with_tail
+    @estimated_length + (@something_has_been_truncated ? tail.bytesize : 0)
+  end
+
+  def single_tag_element?(name)
+    VOID_TAGS.include? name
+  end
+
+  def append_to_truncated_string(string, overridden_length = nil)
+    @truncated_string << string
+    increase_estimated_length(overridden_length || string.bytesize)
+  end
+
+  def attributes_to_string(attributes)
+    return "" if attributes.empty?
+    attributes_string = concatenate_attributes_declaration attributes
+    attributes_string.rstrip
+  end
+
+  def concatenate_attributes_declaration(attributes)
+    attributes.inject(' ') do |string, attribute|
+      key, value = attribute
+      string << "#{key}='#{@html_coder.encode value}' "
+    end
+  end
+
+  def increase_estimated_length(amount)
+    @estimated_length += amount
+    check_max_length_reached
+  end
+
+  def check_max_length_reached
+    @max_length_reached = true if estimated_length_with_tail >= max_length
```

Comment:
what's the reason to keep this as an instance variable rather than just doing the calculation each time it's needed?